flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/SpilledSubpartitionView.java

```diff
@@ -147,7 +147,8 @@ public BufferAndBacklog getNextBuffer() throws IOException, InterruptedException
 			return null;
 		}
 
-		int newBacklog = parent.decreaseBuffersInBacklog(current.isBuffer());
+		parent.decreaseBuffersInBacklog(current.isBuffer());
+		int newBacklog = parent.getBuffersInBacklog();
```

Comment:
Something is wrong here. The line above is synchronized (`decreaseBuffersInBacklog` vs `decreaseBuffersInBacklogUnsafe`) and this `getBuffersInBacklog` is not (note miss-leading lack of `Unsafe` suffix) . Either:

1. the above one doesn't need to be synchronized
2. there is a bug and this one should be synchronized. In that case however, it would be additional unnecessary synchronisation cost, so I think I would prefer to integrate `getBuffersInBacklog` with `decreaseBuffersInBacklog` after all.
3. This doesn't need to be synchronized because of some hacky assumption. In that case either write big comment or probably better and/or safer approach would be to integrate `getBuffersInBacklog` with `decreaseBuffersInBacklog` after all.

Reply:
That is a good question that should be concerned. I considered it again and thought the synchronized is not needed for getting backlog because this value would be final consistent between sender and receiver.

E.g. if the current backlog is 4 after decreasing, the previous behavior would report 4 strictly. The new behavior might report 4 or more than 4 if increasing backlog again before getting. But the result is still correct if reporting 5 because it actually exists. The difference is we report this increase in advance, and the previous behavior would reflect this increase in the next report. The early report might get some extra benefits because the receiver could prepare more credits for it.

It could have two options:
1. Uniform all the ways of getting backlog outside of synchronized.
2. Integrate return backlog in `decreaseBuffersInBacklog` as you mentioned, and the form seems not so bad if adding boolean parameter in `decreaseBuffersInBacklog`. The backlog is strictly returned in this way.

Which option do you prefer?

Reply:
I think that option 1. even if it's correct right now, would be more error prone for the future. Forcing every contributor/committer in this part to rethink this contract/dependencies every time is not a good thing to do.

For example, as you if the backlog increases between those two calls there is no problem, however it takes some time to realize about this fact. Also currently I'm not sure if the backlog can be somehow decreased by a 3rd party (another thread) betwen those two calls and if that happens, if that's ok? Another question, are all of the arithmetic operations on the backlog atomic? 

Long story short, if we can avoid think about the above issues, so much the better - I would pick number 2.

Reply:
I agree with your point of option 2. Based on the current implementation we could confirm no problem, but it might leave risks if changing something in future. It is also easy to understand for others in option 2.